File: src/13/day13.py
# Rock (#) is True
def mirror_positions_col(grid: list[bool], row_length: int, col_length: int) -> int:
    col_keys = [
        sum(grid[i + j * row_length] << j for j in range(col_length))
        for i in range(row_length)
    ]
    first, last = 0, len(col_keys) - 1
    while first < last:
        if col_keys[first] != col_keys[last] or not ((last - first) & 1):
            first += 1
        else:
            f, l = first, last
            palindrome = True
            while f < l and palindrome:
                palindrome = col_keys[f] == col_keys[l]
                f += 1
                l -= 1
            if palindrome:
                return 1 + first + ((last - first) >> 1)
            first += 1
    col_keys.reverse()
    first, last = 0, len(col_keys) - 1
    while first != last:
        if col_keys[first] != col_keys[last] or not ((last - first) & 1):
            first += 1
        else:
            f, l = first, last
            palindrome = True
            iterated = False
            while f < l and palindrome:
                palindrome = col_keys[f] == col_keys[l]
                f += 1
                l -= 1
                iterated = True
            if palindrome:
                return row_length - (1 + first + ((last - first) >> 1))
            first += 1
    return 0

def mirror_positions_row(grid: list[bool], row_length: int, col_length: int) -> int:
    row_keys = [
        sum(grid[i * row_length + j] << j for j in range(row_length))
        for i in range(col_length)
    ]
    first, last = 0, len(row_keys) - 1
    while first < last:
        if row_keys[first] != row_keys[last] or not ((last - first) & 1):
            first += 1
        else:
            f, l = first, last
            palindrome = True
            iterated = False
            while f < l and palindrome:
                palindrome = row_keys[f] == row_keys[l]
                f += 1
                l -= 1
                iterated = True
            if palindrome and iterated:
                return 1 + first + ((last - first) >> 1)
            first += 1
    row_keys.reverse()
    first, last = 0, len(row_keys) - 1
    while first != last:
        if row_keys[first] != row_keys[last] or not ((last - first) & 1):
            first += 1
        else:
            f, l = first, last
            palindrome = True
            iterated = False
            while f < l and palindrome:
                palindrome = row_keys[f] == row_keys[l]
                f += 1
                l -= 1
                iterated = True
            if palindrome:
                return col_length - (1 + first + ((last - first) >> 1))
            first += 1
    return 0

def part1(filename: str) -> int:
    s = 0
    r = 0
    grid = []
    row, col = 0, 0
    with open(filename, 'rt') as f:
        for line in f:
            line = line.strip()
            if line:
                grid.extend(c == '#' for c in line)
                row = len(line)
                col += 1
            else:
                r = mirror_positions_col(grid, row, col)
                if r == 0:
                    r = 100 * mirror_positions_row(grid, row, col)
                row, col = 0, 0
                grid = []
                s += r
    r = mirror_positions_col(grid, row, col)
    if r == 0:
        r = 100 * mirror_positions_row(grid, row, col)
    row, col = 0, 0
    grid = []
    s += r
    #print('col' if odd else 'row', r)
    return s

def part2(filename: str) -> int:
    return 0

if __name__ == '__main__':
    print(part1('day13_input.txt'))
    print(part2('test_input.txt'))

File: src/13/day13.rs
fn mirror_positions(grid_bits: &Vec<u32>, part_two: bool) -> Vec<u32> {
    let mut bits: Vec<u32> = grid_bits.clone();
    let mut results: Vec<u32> = Vec::new();
    for i in 0..2 {
        let mut first: usize = 0;
        let last: usize = bits.len() - 1;
        while first != last {
            if ((last - first) & 1) == 0 {
                first += 1;
            } else {
                let (mut f, mut l) = (first, last);
                let mut palindrome = true;
                let mut count = 0;
                while f < l && palindrome {
                    let bit_diff = bits[f] ^ bits[l];
                    if part_two && bits[f] != bits[l] && (bit_diff & (bit_diff - 1)) == 0 {
                        palindrome = count == 0;
                        count += 1;
                    } else {
                        palindrome = bits[f] == bits[l];
                    }
                    f += 1;
                    l -= 1;
                }
                if palindrome {
                    if i == 0 {
                        results.push((1 + first + ((last - first) >> 1)) as u32);
                    } else {
                        results.push((bits.len() - (1 + first + ((last - first) >> 1))) as u32);
                    }
                }
                first += 1;
            }
        }
        bits.reverse();
    }
    if !part_two && results.len() == 0 {
        results.push(0);
    }
    return results;
}

fn part1(filename: &str) -> u32 {
    let mut s: u32 = 0;
    let mut grid: Vec<bool> = Vec::new();
    let mut row = 0;
    let mut col = 0;
    let lines = std::fs::read_to_string(filename).expect("File not found");
    for line in lines.trim().split('\n').map(|x| x.trim()) {
        if line.len() > 0 {
            grid.extend(line.chars().map(|c| c == '#'));
            row = line.len();
            col += 1;
        } else {
            let col_bits: Vec<u32> = (0..row).map(
                |i|
                (0..col)
                .fold(
                    0,
                    |acc, j|
                    if grid[i + j * row] { acc | (1 << j) } else { acc }
                )
            )
            .collect::<Vec<_>>();
            let mut result = mirror_positions(&col_bits, false)[0];
            if result == 0 {
                let row_bits: Vec<u32> = (0..col).map(
                    |i|
                    (0..row)
                    .fold(
                        0,
                        |acc, j|
                        if grid[i * row + j] { acc | (1 << j) } else { acc }
                    )
                )
                .collect::<Vec<_>>();
                result = 100 * mirror_positions(&row_bits, false)[0];
            }
            row = 0;
            col = 0;
            grid.clear();
            s += result;
        }
    }
    let col_bits: Vec<u32> = (0..row).map(
        |i|
        (0..col)
        .fold(
            0,
            |acc, j|
            if grid[i + j * row] { acc | (1 << j) } else { acc }
        )
    )
    .collect::<Vec<_>>();
    let mut result = mirror_positions(&col_bits, false)[0];
    if result == 0 {
        let row_bits: Vec<u32> = (0..col).map(
            |i|
            (0..row)
            .fold(
                0,
                |acc, j|
                if grid[i * row + j] { acc | (1 << j) } else { acc }
            )
        )
        .collect::<Vec<_>>();
        result = 100 * mirror_positions(&row_bits, false)[0];
    }
    grid.clear();
    s += result;
    return s;
}

fn part2(filename: &str) -> u32 {
    let mut s: u32 = 0;
    let mut grid: Vec<bool> = Vec::new();
    let mut row = 0;
    let mut col = 0;
    let lines = std::fs::read_to_string(filename).expect("File not found");
    for line in lines.trim().split('\n').map(|x| x.trim()) {
        if line.len() > 0 {
            grid.extend(line.chars().map(|c| c == '#'));
            row = line.len();
            col += 1;
        } else {
            let col_bits: Vec<u32> = (0..row).map(
                |i|
                (0..col)
                .fold(
                    0,
                    |acc, j|
                    if grid[i + j * row] { acc | (1 << j) } else { acc }
                )
            )
            .collect::<Vec<_>>();
            let row_bits: Vec<u32> = (0..col).map(
                |i|
                (0..row)
                .fold(
                    0,
                    |acc, j|
                    if grid[i * row + j] { acc | (1 << j) } else { acc }
                )
            )
            .collect::<Vec<_>>();
            let mut part_one_result = mirror_positions(&col_bits, false)[0];
            if part_one_result == 0 {
                part_one_result = 100 * mirror_positions(&row_bits, false)[0];
            }
            let mut results = mirror_positions(&col_bits, true);
            results.extend(mirror_positions(&row_bits, true).into_iter().map(|x| 100*x));
            if results.len() == 0 {
                results.push(0);
            }
            row = 0;
            col = 0;
            grid.clear();
            s += results
                .iter()
                .filter(|&&x| x != part_one_result)
                .reduce(|x1, x2| if x1 < x2 { x1 } else { x2 })
                .unwrap();
        }
    }
    let col_bits: Vec<u32> = (0..row).map(
        |i|
        (0..col)
        .fold(
            0,
            |acc, j|
            if grid[i + j * row] { acc | (1 << j) } else { acc }
        )
    )
    .collect::<Vec<_>>();
    let row_bits: Vec<u32> = (0..col).map(
        |i|
        (0..row)
        .fold(
            0,
            |acc, j|
            if grid[i * row + j] { acc | (1 << j) } else { acc }
        )
    )
    .collect::<Vec<_>>();
    let mut part_one_result = mirror_positions(&col_bits, false)[0];
    if part_one_result == 0 {
        part_one_result = 100 * mirror_positions(&row_bits, false)[0];
    }
    let mut results = mirror_positions(&col_bits, true);
    results.extend(mirror_positions(&row_bits, true).into_iter().map(|x| 100*x));
    if results.len() == 0 {
        results.push(0);
    }
    grid.clear();
    s += results
        .iter()
        .filter(|&&x| x != part_one_result)
        .reduce(|x1, x2| if x1 < x2 { x1 } else { x2 })
        .unwrap();
    return s;
}

fn main() {
    println!("{}", part1("day13_input.txt"));
    println!("{}", part2("day13_input.txt"));
}